File: Arrays/Easy/Palindrome/Palindrome.cpp
// Copyrights to venkys.io
// For more programs visit venkys.io 
// CPP program for Palindrome

// time complexity:O(n)
// space complexity:O(1)

#include <iostream>
#include <cctype>
using namespace std;

// Function to check if a string is a palindrome
bool checkPalindrome(string str) {
    int len = str.size();
    int i = 0, j = len - 1;

    while (i < j) {
        // Ignore non-alphanumeric characters
        // Ignore non-alphanumeric characters from the beginning of the string
        while (i < len && !isalnum(str[i])) {
            i++;
        }

        // Ignore non-alphanumeric characters from the end of the string
        while (j >= 0 && !isalnum(str[j])) {
            j--;
        }

        // If characters at positions i and j are not equal, the string is not a palindrome
        if (i < j) {
            // Compare characters case-insensitively
            if (tolower(str[i]) != tolower(str[j])) {
                return false;
            }
            // Move the pointers towards the center
            i++;
            j--;
        }
    }
    // If the loop completes without returning false, the string is a palindrome
    return true;
}

int main() {

    string str;

    // Get input from the user
    // cout << "Enter a string: ";
    getline(cin, str);

    // Check if the string is a palindrome using the checkPalindrome function
    if (checkPalindrome(str)) {
        cout << "True"<< endl;
    } else {
        cout << "False" << endl;
    }

    return 0;
}

File: Arrays/Easy/Palindrome/Palindrome.java
// Copyrights to venkys.io
// For more programs visit venkys.io 
// Java program for Palindrom

// time complexity:O(n)
// space complexity:O(1)

import java.util.Scanner;

public class Main {
    // Function to compute reverse of given string and return boolean value by matching it against the original string
    public static boolean isPalindrome(String s) {
        // Remove non-alphanumeric characters
        String cleanedString = s.replaceAll("[^a-zA-Z0-9]", "");
        // Declare an empty string to store the reversed version
        String rev = "";
        // Loop to iterate through the characters of the string in reverse order
        for (int i = cleanedString.length() - 1; i >= 0; i--) {
            // Concatenate each character to the reversed string
            rev += cleanedString.charAt(i);
        }
        // Check if the reversed string is equal to the original string
        return rev.equals(cleanedString);
    }

    public static void main(String args[]) {
        // Scanner to take user input
        Scanner scanner = new Scanner(System.in);

        // System.out.print("Enter a string: ");
        // Read user input
        String s = scanner.nextLine();

        // Calling the palindrome-checking method and storing the result in a boolean variable
        boolean isPal = isPalindrome(s);

        if (isPal) {
            System.out.println("True");
        } else {
            System.out.println("False");
        }

        // Close the scanner
        scanner.close();
    }
}
